godot-core/src/meta/property_info.rs

```diff
@@ -179,6 +180,48 @@ impl PropertyInfo {
         }
     }
 
+    /// Consumes self, moving the values into `sys::GDExtensionPropertyInfo` pointer.
+    ///
+    /// # Safety
+    ///
+    /// * `property_info_ptr` must be valid.
+    pub(crate) unsafe fn move_into_property_info_ptr(
+        self,
+        property_info_ptr: *mut sys::GDExtensionPropertyInfo,
+    ) {
+        let class_name = if self.class_name == ClassName::none() {
+            (*property_info_ptr).class_name
+        } else {
+            sys::SysPtr::force_mut(self.class_name.string_sys())
+        };
+        property_info_ptr.write(self.property_sys());
+        (*property_info_ptr).class_name.write(*class_name);
+
+        // Decreasing the refcount on refcounted types (such as StringName or GString)
+        // is a responsibility of the `property_info_ptr`'s owner.
+        std::mem::forget(self);
```

Comment:
With the current implementation, Godot cannot know when to decrease the refcounts though -- you simply overwrite the pointers. So I think this would leak memory.

Reply:
I'm not sure about that – `new_from_string_sys` doesn't increase the refcount, we also want values created in user space to persist and I haven't noticed the memory leak with previous implementation.

I'll validate it with valgrind and Godot compiled with sanitizers support: <https://docs.godotengine.org/en/stable/contributing/development/debugging/using_sanitizers.html> (same we use for memchecks afaik)

Reply:
I _think_ with the [above approach](https://github.com/godot-rust/gdext/pull/1030#discussion_r1947840666) you can move the dynamically-allocated objects back to Godot without leaking memory.

I'm then not sure whether `forget` is necessary -- rustc might complain if some fields have been moved out, and it can no longer invoke `Drop::drop(&mut self)`.

Reply:
> I'm not sure about that – `new_from_string_sys` doesn't increase the refcount, we also want values created in user space to persist and I haven't noticed the memory leak with previous implementation.

`new_from_string_sys` absolutely does increase the ref count. that's the point of it, if you dont want to increase the refcount you'd use `borrow_string_sys`.